app/services/authorization/user_permissible_service.rb

```diff
@@ -0,0 +1,138 @@
+module Authorization
+  class UserPermissibleService
+    attr_accessor :user
+
+    def initialize(user)
+      @user = user
+      @cached_permissions = Hash.new do |hash, key|
+        hash[key] = Set.new
+      end
+    end
+
+    def allowed_globally?(permission)
+      perms = Authorization.contextual_permissions(permission, :global, raise_on_unknown: true)
+      return true if admin_and_all_granted_to_admin?(perms)
+
+      cached_permissions(:global).intersect?(perms.map(&:name))
+    end
+
+    def allowed_in_project?(permission, projects_to_check)
+      perms = Authorization.contextual_permissions(permission, :project, raise_on_unknown: true)
+      return false if projects_to_check.blank?
+      return false unless authorizable_user?
+
+      projects = Array(projects_to_check)
+
+      projects.all? do |project|
+        allowed_in_single_project?(perms, project)
+      end
+    end
+
+    def allowed_in_any_project?(permission)
+      perms = Authorization.contextual_permissions(permission, :project, raise_on_unknown: true)
+      return true if admin_and_all_granted_to_admin?(perms)
+
+      Project.allowed_to(user, perms).exists?
+    end
+
+    def allowed_in_entity?(permission, entities_to_check, entity_class)
+      return false if entities_to_check.blank?
+      return false unless authorizable_user?
+
+      context = entity_class.model_name.element.to_sym
+      perms = Authorization.contextual_permissions(permission, context, raise_on_unknown: true)
+
+      entities = Array(entities_to_check)
+
+      entities.all? do |entity|
+        allowed_in_single_entity?(perms, entity)
+      end
+    end
+
+    def allowed_in_any_entity?(permission, entity_class, in_project: nil)
+      context = entity_class.model_name.element.to_sym
+      perms = Authorization.contextual_permissions(permission, context, raise_on_unknown: true)
+      return true if admin_and_all_granted_to_admin?(perms)
+
+      if in_project
+        WorkPackage.allowed_to(user, perms).exists?(project: in_project)
+      else
+        WorkPackage.allowed_to(user, perms).exists?
+      end
+    end
```

Comment:
Shouldn't it be using `entity_class` instead of assuming `WorkPackage`?

Reply:
Absolutely!